moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp

```diff
@@ -132,61 +118,129 @@ bool RuckigSmoothing::applySmoothing(robot_trajectory::RobotTrajectory& trajecto
       {
-        ruckig_input.target_velocity.at(joint) *= 0.9;
-        // Propogate the change in velocity to position and acceleration, too.
-        // This derives from a first-order approximation, e.g. x_i+1 = x_i + v_i * delta_t
-        // and 0.9 * x_i+1 = 0.9 * (x_i + v_i * delta_t)
-        ruckig_input.target_position.at(joint) -= (0.9 * 0.9 - 1) * timestep * ruckig_input.target_velocity.at(joint);
-        ruckig_input.target_acceleration.at(joint) =
-            (ruckig_input.target_velocity.at(joint) - ruckig_output.new_velocity.at(joint)) / timestep;
+        next_waypoint->setVariablePosition(idx.at(joint), ruckig_output.new_position.at(joint));
+        next_waypoint->setVariableVelocity(idx.at(joint), ruckig_output.new_velocity.at(joint));
+        next_waypoint->setVariableAcceleration(idx.at(joint), ruckig_output.new_acceleration.at(joint));
       }
-      velocity_magnitude = getTargetVelocityMagnitude(ruckig_input, num_dof);
-      // Run Ruckig
-      ruckig_result = ruckig.update(ruckig_input, ruckig_output);
-      // check for backward motion
-      backward_motion_detected = checkForLaggingMotion(num_dof, ruckig_input, ruckig_output);
+      next_waypoint->update();
     }
 
-    if (ruckig_result != ruckig::Result::Working)
+    // If ruckig failed, the duration of the seed trajectory likely wasn't long enough.
+    // Try duration extension several times.
+    if (ruckig_result == ruckig::Result::Working)
     {
-      RCLCPP_ERROR_STREAM(LOGGER, "Ruckig trajectory smoothing failed at waypoint " << waypoint_idx);
-      RCLCPP_ERROR_STREAM(LOGGER, "Ruckig error: " << ruckig_result);
-      return false;
+      smoothing_complete = true;
     }
-
-    // Overwrite pos/vel/accel of the target waypoint
-    for (size_t joint = 0; joint < num_dof; ++joint)
+    else
     {
-      next_waypoint->setVariablePosition(idx.at(joint), ruckig_output.new_position.at(joint));
-      next_waypoint->setVariableVelocity(idx.at(joint), ruckig_output.new_velocity.at(joint));
-      next_waypoint->setVariableAcceleration(idx.at(joint), ruckig_output.new_acceleration.at(joint));
+      // If Ruckig failed, it's likely because the original seed trajectory did not have a long enough duration when
+      // jerk is taken into account. Extend the duration and try again.
+      initializeRuckigState(ruckig_input, ruckig_output, *trajectory.getFirstWayPointPtr(), num_dof, idx);
+      duration_extension_factor *= DURATION_EXTENSION_FRACTION;
+      for (size_t waypoint_idx = 1; waypoint_idx < num_waypoints; ++waypoint_idx)
+      {
+        trajectory.setWayPointDurationFromPrevious(
+            waypoint_idx, DURATION_EXTENSION_FRACTION * trajectory.getWayPointDurationFromPrevious(waypoint_idx));
+        // TODO(andyz): re-calculate waypoint velocity and acceleration here?
+      }
+
+      timestep = trajectory.getAverageSegmentDuration();
+      ruckig_ptr.reset(new ruckig::Ruckig<0>{ num_dof, timestep });
```

Comment:
Why? I think it's more semantically clear to do a `reset()` here since the pointer was initialized to something else before. (Unlike your previous comment, where the pointer hadn't been initialized yet.)

Reply:
This is the primary reason I generally encourage the make functions over reset and new with the constructor: https://stackoverflow.com/a/41001856

It seems that as of C++17 that reason doesn't matter much.

The second reason is that when you call make functions it only does one dynamic memory allocation (control block + new object).  When you call new in a smart pointers constructor it does two memory allocations (your call to new, then inside the constructor did the smart pointer a second one for the control block).  When you call reset a similar thing happens because it can't reuse the control block it has because someone else might also hold a reference to it.  The short answer is that using reset or the constructor with new is a pessimization (double the amount of calls to malloc and more fragmented memory) vs using the make functions.